File: 71_simplify_path/main.java
import java.util.Stack;
import java.util.Arrays;

class Solution {
    public String simplifyPath(String path) {
        Stack<String> stack = new Stack<>();

        String[] tokens = Arrays.stream(path.split("/"))
                        .filter(x -> !x.isEmpty())
                        .toArray(String[]::new);

        for (int i = 0; i < tokens.length; ++i) {
            if (tokens[i].equals("..")) {
                if (stack.size() > 0) {
                    stack.pop();
                }
            } else if (tokens[i].equals(".")) {
                // nothing
            } else {
                stack.push(tokens[i]);
            }
        }

        return "/" + String.join("/", stack.toArray(new String[stack.size()]));
    }
}

File: 71_simplify_path/main.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        parts = path.split("/")

        stack = []
        for p in parts:
            if len(p) == 0:
                continue
            elif p == ".":
                continue
            elif p == '..':
                if stack:
                    stack.pop()
            else:
                stack.append(p)
        
        return "/" + "/".join(stack)
            
        
